### ISSUE TITLE
Extension of ProteinBERT with new OutputType for PWM prediction

### ISSUE THREAD

**Issue by Debbie1812** [nadavbra/protein_bert#106]
Hello,
I am trying to extend ProteinBERT to use it for predicting transcription factor motifs as position weight matrices (pwms) from transcription factor dna binding domain sequences (amino acid sequences).
To do so, i extended the output types with
```
def __init__(self, is_seq, output_type):
        self.is_seq = is_seq
        self.output_type = output_type
        self.is_numeric = (output_type == 'numeric')
        self.is_binary = (output_type == 'binary')
        self.is_categorical = (output_type == 'categorical')
        self.is_distribution = (output_type == 'distribution')  # new output_type for PWM prediction
```
My aim is to predict pwms of size 26*4 (26 positions). I have problems with understanding how I have to adapt the ```encode_Y()``` function in the finetuning. I want wo try the prediction with ```is_seq=True``` at the model definition to use the ```pretraining_output_seq_layer``` as ```last_hidden_layer```.
I am currently stuck at:
``` 
def encode_Y(raw_Y, output_spec, seq_len = 512):
    if output_spec.output_type.is_distribution:   # new for PWM prediction
        if output_spec.output_type.is_seq:
            ...
        else:    
            return np.stack(raw_Y.values), np.ones(len(raw_Y))
```
My idea was, that I have amino acid sequences of length 128 (padded) and I always want to predict PWMs of size 26*4 (also padded with uniform distribution at both ends).
How must I adapt the ```encode_Y()``` function? What output format does it expect?

I hope this question was understandable.
Best regards 
Debbie

**Comment by nadavbra** [comment]
@Debbie1812 Can you clarify why you want to adapt the encode_Y function? Why not use one of the existing types supported by ProteinBERT?

**Comment by Debbie1812** [comment]
Yes okay. I want to predict position weight *matrices* ([A,C,G,T],[0.3, 0.3, 0.2, 0.2], ...) so I also need another loss function and metric to compare the true and the predicted matrix. None of the existing Output Types seemed to fit, why i decided to add a new type ```distribution``` . For example I also had to adapt the output layer in ```create_model()```:
```
elif self.output_spec.output_type.is_distribution:  
            dens = keras.layers.Dense(104, activation = None, kernel_initializer='glorot_normal')(last_hidden_layer)
            pwm_shape = keras.layers.Reshape((26, 4))(dens)
            output_layer = keras.layers.Softmax(axis=-1)(pwm_shape)
            loss = ...
            metrics = ...
```
 ( I am still working on this)

Now, my problem is I wanted to use the local representations (```is_seq=True```) but this gives me an error in the ```finetune()```method, when ```encode_seq_Y()``` is called:
```
File c:\protein_bert\proteinbert\finetuning.py:210, in encode_Y(raw_Y, output_spec, seq_len)
--> [210] return encode_seq_Y(raw_Y, seq_len, output_spec.output_type.is_binary, output_spec.unique_labels)
    211     elif output_spec.output_type.is_categorical:
    212         return encode_categorical_Y(raw_Y, output_spec.unique_labels), np.ones(len(raw_Y))

File c:\protein_bert\proteinbert\finetuning.py:220, in encode_seq_Y(seqs, seq_len, is_binary, unique_labels)
    218 def encode_seq_Y(seqs, seq_len, is_binary, unique_labels):
--> [220] label_to_index = {str(label): i for i, label in enumerate(unique_labels)}
    222     Y = np.zeros((len(seqs), seq_len), dtype = int)
    223     sample_weigths = np.zeros((len(seqs), seq_len))
```
I am not sure, how I have to adapt the encoding of Y (matrices) for this problem.
